File: Chapter01/javascript/flocks.js
// Naming here should be more descriptive, but I'm following the book's example

const conjoin = (flockX, flockY) => flockX + flockY;
const breed = (flockX, flockY) => flockX * flockY;
/*
  * Case where the other flock raids the first one, kills the leader
  * and takes others as prisoners.
*/
const raidFlock = (flockX, flockY) => {
  const newFlockY = flockY
  const newFlockX = flockX + newFlockY - 1;
  return [newFlockX, 0]
}

const flockX = 4;
const flockY = 2;

const result = raidFlock(breed(flockX, conjoin(flockX, flockY)), breed(flockX, flockY));
console.log(`Result: [${result}]`);

File: Chapter01/haskell/flocks.hs

conjoin :: Integer -> Integer -> Integer
conjoin flock_x flock_y = flock_x + flock_y

breed :: Integer -> Integer -> Integer
breed flock_x flock_y = flock_x * flock_y

raid_flock :: Integer -> Integer -> [Integer]
raid_flock flock_x flock_y = 
  [conjoin flock_x flock_y - 1, 0]


main :: IO ()
main = do 
  let 
    flock_x = 4
    flock_y = 2
  
  let result = do raid_flock (breed flock_x (conjoin flock_x flock_y)) (breed flock_x flock_y)

  print result